File: packages/ogdf.js/src/module/config/MultilevelBuilder.js
import createModule from '../module-helper'
import { PARAMETER_TYPE } from '../../utils/parameter-type'
const NAME = 'MultilevelBuilder'

const MODULE_DIRECTORY = {
    EdgeCoverMerger: {
        edgeLengthAdjustment: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 0
        },
        factor: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 2.0
        }
    },
    IndependentSetMerger: {
        edgeLengthAdjustment: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 0
        },
        searchDepthBase: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 2.0
        }
    },
    LocalBiconnectedMerger: {
        edgeLengthAdjustment: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 0
        },
        factor: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 2.0
        }
    },
    MatchingMerger: {
        edgeLengthAdjustment: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 0
        },
        selectByNodeMass: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        }
    },
    RandomMerger: {
        edgeLengthAdjustment: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 0
        },
        factor: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 2.0
        }
    },
    SolarMerger: {
        edgeLengthAdjustment: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 0
        }
    }
}

const MultilevelBuilder = createModule(NAME, MODULE_DIRECTORY)
export default MultilevelBuilder

File: packages/ogdf.js/src/module/src/MultilevelBuilder.cpp
#include "../main.h"

#include <ogdf/energybased/multilevel_mixer/EdgeCoverMerger.h>
#include <ogdf/energybased/multilevel_mixer/IndependentSetMerger.h>
#include <ogdf/energybased/multilevel_mixer/LocalBiconnectedMerger.h>
#include <ogdf/energybased/multilevel_mixer/MatchingMerger.h>
#include <ogdf/energybased/multilevel_mixer/RandomMerger.h>
#include <ogdf/energybased/multilevel_mixer/SolarMerger.h>

EM_PORT_API(EdgeCoverMerger *)
MultilevelBuilder_EdgeCoverMerger(int edgeLengthAdjustment, double factor)
{
    EdgeCoverMerger *merger = new EdgeCoverMerger();
    merger->setEdgeLengthAdjustment(edgeLengthAdjustment);
    merger->setFactor(factor);
    return merger;
}

EM_PORT_API(IndependentSetMerger *)
MultilevelBuilder_IndependentSetMerger(int edgeLengthAdjustment, float searchDepthBase)
{
    IndependentSetMerger *merger = new IndependentSetMerger();
    merger->setEdgeLengthAdjustment(edgeLengthAdjustment);
    merger->setSearchDepthBase(searchDepthBase);
    return merger;
}

EM_PORT_API(LocalBiconnectedMerger *)
MultilevelBuilder_LocalBiconnectedMerger(int edgeLengthAdjustment, double factor)
{
    LocalBiconnectedMerger *merger = new LocalBiconnectedMerger();
    merger->setEdgeLengthAdjustment(edgeLengthAdjustment);
    merger->setFactor(factor);
    return merger;
}

EM_PORT_API(MatchingMerger *)
MultilevelBuilder_MatchingMerger(int edgeLengthAdjustment, bool selectByNodeMass)
{
    MatchingMerger *merger = new MatchingMerger();
    merger->setEdgeLengthAdjustment(edgeLengthAdjustment);
    merger->selectByNodeMass(selectByNodeMass);
    return merger;
}

EM_PORT_API(RandomMerger *)
MultilevelBuilder_RandomMerger(int edgeLengthAdjustment, double factor)
{
    RandomMerger *merger = new RandomMerger();
    merger->setEdgeLengthAdjustment(edgeLengthAdjustment);
    merger->setFactor(factor);
    return merger;
}

EM_PORT_API(SolarMerger *)
MultilevelBuilder_SolarMerger(int edgeLengthAdjustment)
{
    SolarMerger *merger = new SolarMerger();
    merger->setEdgeLengthAdjustment(edgeLengthAdjustment);
    return merger;
}
